src/sentry/services/hybrid_cloud/filter_query.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import abc
+from typing import TYPE_CHECKING, Any, Callable, Generic, List, Optional, Protocol, TypeVar
+
+from django.db.models import QuerySet
+
+if TYPE_CHECKING:
+    from sentry.api.serializers import Serializer
+    from sentry.services.hybrid_cloud.auth import AuthenticationContext
+    from sentry.services.hybrid_cloud.user import APIUser
+
+
+class AsSerializer(Protocol):
+    def as_serializer(self) -> Serializer:
+        pass
+
+
+FILTER_ARGS = TypeVar("FILTER_ARGS")  # A typedict
+RPC_RESPONSE = TypeVar("RPC_RESPONSE")
+SERIALIZER_ENUM = TypeVar("SERIALIZER_ENUM", bound=AsSerializer)
+BASE_MODEL = TypeVar("BASE_MODEL")
+
+# In the future, this ought to be a pass through type that does not get double serializer, and which cannot be
+# inspected by code.
+OpaqueSerializedResponse = Any
+
+
+# A collection of classes that can be used to quickly provision interfaces & implementations for basic RPC Hybrid Cloud service implementations.
+# It enforces several niceties:
+#   * Type safety at call sites: the public API can be inferred, unknown arguments are rejected,
+#     all filter arguments are optional, and nullable filters are explicitly Optional[...]
+#   * Filter argument validation: an arbitrary function can be used to ensure that proper filters are applied
+#     e.g. we don't want to just query for every row, so we can enforce that at least one filter key is provided
+#   * A standard interface across most of our simple rpc services
+
+# This class can be inherited from to add this functionality to the public interface for a service.
+# E.g. class UserService(FilterQueryInterface[
+#        UserQueryArgs, APIUser, Union[UserSerializerResponse, UserSerializerResponseSelf]
+#      ], InterfaceWithLifecycle):
+#         ...
+class FilterQueryInterface(Generic[FILTER_ARGS, RPC_RESPONSE, SERIALIZER_ENUM], abc.ABC):
```

Comment:
This is really the part that bothers me the most.  Python generics are, useful, but they are not nearly as ergonomic as they are in other languages.  For instance,  creating a new service and understanding how to fill all these things out, understanding how to setup the appropriate typedict and dataclasses, what the limitations and expectations of they function -- it's not captured well here in the types, and it can be hard (no way to bound= TypedDict or dataclasses).  So the goal of a consistent interface is only partially solved -- people don't have to agree on what sorts of things these different type holes become.